### PR TITLE
Add `width` and `height` option

### PR DESCRIPTION
**Pull request by piranna**
This pull-request add support for optional `width` and `height` options, auto-adjusting size if any (or both) of them is not defined, also when running in Windows. It also unify the `file` and `buffer` functions, and fix two bugs related to alpha channel and when resized image has an odd number of rows.

### CODE REVIEW

**Comment by piranna** [comment]
This is open to comments and discussion.

**Comment by goloroden** [comment]
This is great work, thanks a lot for it 👍 

One thing: In the `README.md` it says `heigh` instead of `height`.

**Comment by piranna** [comment]
> This is great work, thanks a lot for it 
> 
> One thing: In the README.md it says heigh instead of height.

You are welcome ;-) Typo fixed, and also fixed some special cases when only one of the half blocks is full transparent and optimized the generation of escape codes, although this last point can be improved a bit more.

File: readme.md [hunk 1]
```diff
@@ -22,40 +23,43 @@ $ npm install terminal-image
 const terminalImage = require('terminal-image');
 
 (async () => {
-	console.log(await terminalImage.file('unicorn.jpg'));
+	console.log(await terminalImage('unicorn.jpg'));
 })();
 ```
 
+Optionally, you can specify the `height` and/or `width` to scale the image. To
+maintain aspect ratio, define only one of them.
```
Comment (sindresorhus) [inline readme.md]: We should always maintain aspect ratio. The image should just fit inside `width` and `height`.

File: readme.md [hunk 2]
```diff
@@ -22,40 +23,43 @@ $ npm install terminal-image
...
-#### imageBuffer
 
-Type: `Buffer`
+## API
 
-Buffer with the image.
+Supports PNG and JPEG images.
 
-### terminalImage.file(filePath)
+### terminalImage(fileBuffer, [options])
```
Comment (sindresorhus) [inline readme.md]: `fileBuffer` is not correct when it can also be a filePath. Rather just call it `input`.

File: readme.md [hunk 3]
```diff
@@ -22,40 +23,43 @@ $ npm install terminal-image
...
 
 
 ## Related
 
-- [terminal-image-cli](https://github.com/sindresorhus/terminal-image-cli) - CLI for this module
-- [terminal-link](https://github.com/sindresorhus/terminal-link) - Create clickable links in the terminal
+- [terminal-image-cli](https://github.com/sindresorhus/terminal-image-cli) - CLI
+	for this module
+- [terminal-link](https://github.com/sindresorhus/terminal-link) - Create
+	clickable links in the terminal
```
Comment (sindresorhus) [inline readme.md]: Don't do unrelated changes.

File: index.js [hunk 4]
```diff
@@ -1,47 +1,163 @@
...
+	} else {
+		width = bitmap.width / terminalCharWidth * height / bitmap.height;
+	}
+
+	// Each character has two vertical blocks, so we double the number of rows
+	height *= 2;
 
-async function render(buffer) {
-	const image = await Jimp.read(buffer);
-	const columns = process.stdout.columns || 80;
```
Comment (sindresorhus) [inline index.js]: The image should never be wider than the terminal. That makes it just a mess of colors.

screen shot 2018-05-16 at 23 39 27

File: index.js [hunk 5]
```diff
@@ -1,47 +1,163 @@
 'use strict';
-const util = require('util');
-const fs = require('fs');
-const chalk = require('chalk');
+
 const Jimp = require('@sindresorhus/jimp');
-const termImg = require('term-img');
+const chalk = require('chalk');
+const {string: termImg} = require('term-img');
+const terminalCharWidth = require('terminal-char-width');
```
Comment (sindresorhus) [inline index.js]: I don't trust this module. Look at its source. And I don't understand why it's used in the first place.

File: index.js [hunk 6]
```diff
@@ -1,47 +1,163 @@
...
+}
+
+function getFgColor({r, g, b}) {
+	return chalk.rgb(r, g, b);
+}
+
+function getGlyph(upper, lower) {
+	if (upper.a) {
+		if (lower && equalColor(upper, lower)) {
+			return '█';
```
Comment (sindresorhus) [inline index.js]: Is it really worth all this complication just to save some glyphs? Like, what are the benefits of using a whole glyph in this scenario?

File: index.js (hunk 4 above)
Comment (sindresorhus) [reply index.js]: By default, it should work like before and just fill the terminal while preserving aspect ratio.

File: index.js [hunk 7]
```diff
@@ -52,7 +52,7 @@ function getGlyph(upper, lower) {
 	return ' ';
 }
 
-async function render(fileBuffer, {height, width}) {
+async function render(fileBuffer, {asArray, height, width}) {
```
Comment (sindresorhus) [inline index.js]: Why? What's the use-case for this?

**Comment by sindresorhus** [comment]
What do you think about supporting setting width/height as percentage (string) too? So you could do `{width: '50%', height: '50%'}` to make the image take up max 50% of the width and the height, while still preserving aspect ratio. That would solve the earlier `factor` need, but would be more flexible.

File: readme.md (hunk 1 above)
Comment (piranna) [reply readme.md]: It depends of the application, but I agree it's the desired behaviour for most of the people. What about a flag enabled by default to do the fitting when both are set?

File: readme.md (hunk 2 above)
Comment (piranna) [reply readme.md]: Nice name! I was not able to find a good one :+1:

File: index.js (hunk 4 above)
Comment (piranna) [reply index.js]: Ok :+1:

File: index.js (hunk 5 above)
Comment (piranna) [reply index.js]: I've seen the source code and didn't find anything strange... Can you explain your concerns? I'm using it to get the width/height relationship of the terminal characters (0.725 on Windows, 0.5 on everything else, like GNU Unifont) so the aspect ration can be preserverd.

File: index.js (hunk 6 above)
Comment (piranna) [reply index.js]: Both the upper and lower half blocks has the same color, so it can be defined with a full block glyph and the foreground color and the background don't need to change. This complication allow use less ANSI escape codes. There's the problem `chalk` add reset ANSI codes at the end that are mostly useless since we are going the set the color again later, so another change I was thinking is to use directly the `ansi-styles` module to do the color transitions, but this can be done in a different pull-request.

**Comment by piranna** [comment]
> What do you think about supporting setting width/height as percentage (string) too? So you could do {width: '50%', height: '50%'} to make the image take up max 50% of the width and the height, while still preserving aspect ratio. That would solve the earlier factor need, but would be more flexible.

I like this idea, problem is, percentage of what? :-P We need the sizes of the parent element, being this the screen or whatever. That's the same reason why I didn't bounded the size to the screen sizes by default, for directly drawing to the screen it's ok, but if you want to convert the image to a string, the screen size doesn't make sense and if you want to limit the size to it I think you should set the dimensions explicitly... what do you think?

File: readme.md (hunk 1 above)
Comment (sindresorhus) [reply readme.md]: I'm not opposed to an option for it if you can elaborate on use-cases for it?

File: index.js (hunk 6 above)
Comment (sindresorhus) [reply index.js]: 👍 Alright

**Comment by goloroden** [comment]
Is there a chance that this gets merged even if it's not absolutely perfect?

Because, work on this started a month ago, and the initial solution was fine. And then another feature was added, and even another one, and so everyone is still waiting for this PR 😉 

I mean, the initial solution was not perfect as well, so even if this gets merged in the current state, it would be a dramatic improvement.

What do you think?

File: readme.md (hunk 1 above)
Comment (sindresorhus) [reply readme.md]: Alright, let's go with a flag for it: `preserveAspectRatio: true` by default.

**Comment by sindresorhus** [comment]
> Is there a chance that this gets merged even if it's not absolutely perfect?

It's not about being perfect, but if we don't get the API right now, we'll have to do another set of breaking changes later on.

**Comment by sindresorhus** [comment]
I'm sure @piranna is just busy and will be back here eventually to finish it up. In the meantime, you can just depend on his fork.

**Comment by sindresorhus** [comment]
@piranna Still interested in finishing this up?

**Comment by sindresorhus** [comment]
Closing this in favor of #14
